File: chain_of_responsibility.js
let Request = function (amount) {
    this.amount = amount;
    console.log("Requested: $" + amount)
};

Request.prototype = {
    get: function(bill) {
        let count = Math.floor(this.amount / bill);
        this.amount -= count * bill;
        console.log("Dispense " + count + " $" + bill + " bills");
        return this;
    }
};

let main = () => {
	let request = new Request(378);
	request.get(100).get(50).get(20).get(10).get(5).get(1);
}

main()

File: chain_of_responsibility.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct Request{
	int amount;
	void (*get) (struct Request* req, int a);
} Request;

void getfunc (Request* request, int bill){
	int count = floor(request->amount/bill);
	request->amount = request->amount - (count * bill);
	printf("Dispense %d $%d bills\n", count, bill);
}

struct Request* init (int start){
	Request *req = malloc(sizeof(Request));
	printf("Requested: $%d\n", start);
	req->amount = start;
	req->get = getfunc;
	return req;
}

int main(int argc, char const *argv[]){
	Request* request = init(378);
	request->get(request,100);
	request->get(request,50);
	request->get(request,20);
	request->get(request,10);
	request->get(request,5);
	request->get(request,1);
	free(request);
	return 0;
}
